server/src/lib/model/achievements.ts

```diff
@@ -85,44 +88,40 @@ const bonus_condition_sql = {
     `,
 };
 const db = getMySQLInstance();
-export default {
-  async earnBonus(type: AchievementType, args: any[]) {
-    let earned = false;
 
-    const [[res]] = await db.query(bonus_condition_sql[type], args);
-    const { win_bonus, bonus_winner } = res || {
-      win_bonus: false,
-      bonus_winner: '',
-    };
-    if (win_bonus && bonus_winner) {
-      const ret = await db.query(
-        `
-              INSERT INTO earn (achievement_id, client_id) VALUES ((SELECT id FROM achievements WHERE name = ?), ?)
-            `,
-        [type, bonus_winner]
-      );
-      earned = ret && ret[0] && ret[0].affectedRows > 0 ? true : false;
-    }
-    return earned;
-  },
+export const earnBonus = async (type: AchievementType, args: any[]) => {
+  let earned = false;
 
-  async hasEarnedBonus(
-    type: AchievementType,
-    client_id: string,
-    challenge: ChallengeToken
-  ) {
-    const [[{ earned }]] = await db.query(
+  const [[res]] = await db.query(bonus_condition_sql[type], args);
+  const { win_bonus = false, bonus_winner = '' } = res || {};
+  if (win_bonus && bonus_winner) {
+    const ret = await db.query(
       `
-        SELECT earn.client_id IS NOT NULL AS earned
-        FROM challenges
-        LEFT JOIN achievements ON challenges.id = achievements.challenge_id AND achievements.name = ?
-        LEFT JOIN earn ON achievements.id = earn.achievement_id
-            AND earn.client_id = ?
-            AND earn.earned_at BETWEEN start_date AND TIMESTAMPADD(WEEK, 3, start_date)
-        WHERE challenges.url_token = ?
-        `,
-      [type, client_id, challenge]
+            INSERT INTO earn (achievement_id, client_id) VALUES ((SELECT id FROM achievements WHERE name = ?), ?)
```

Comment:
Simplicity had taken my mind back then, but now, I think it could be better to add a LEFT JOIN to achievement id sub-query.
Because if different challenges have different achievements, only name is not enough to identify an achievement, together with challenge_id will do.